compiler/plugins/target/ROCM/Dialect/ROCM/Transforms/ApplyBuiltinPDLPatterns.cpp

```diff
@@ -78,6 +78,55 @@ static LogicalResult matchContraction(PatternRewriter &rewriter,
   return success();
 }
 
+/// PDL utility that checks whether the size of the dimension of the provided
+/// type is a multiple of the divisor.
+static LogicalResult dimIsMultipleOf(PatternRewriter &rewriter, Type type,
+                                     Attribute dim, Attribute divisor) {
+  auto shapedType = dyn_cast<mlir::ShapedType>(type);
+  if (!shapedType) {
+    return failure();
+  }
+  auto dimInt = dyn_cast<IntegerAttr>(dim);
+  if (!dim) {
+    return failure();
+  }
+  auto divisorInt = dyn_cast<IntegerAttr>(divisor);
+  if (!divisor) {
+    return failure();
+  }
+  return success(shapedType.getShape()[dimInt.getInt()] % divisorInt.getInt() ==
+                 0);
+}
+
+/// PDL utility that checks whether the size of the dimension of the provided
+/// type is within the specified range.
+static LogicalResult dimIsBound(PatternRewriter &rewriter, Type type,
+                                Attribute dim, Attribute lowerBound,
+                                Attribute upperBound) {
+  auto shapedType = dyn_cast<mlir::ShapedType>(type);
+  if (!shapedType) {
+    return failure();
+  }
+  auto dimInt = dyn_cast<IntegerAttr>(dim);
+  if (!dimInt) {
+    return failure();
+  }
+  auto lowerBoundInt = dyn_cast<IntegerAttr>(lowerBound);
+  auto upperBoundInt = dyn_cast<IntegerAttr>(upperBound);
+  const int64_t size = shapedType.getShape()[dimInt.getInt()];
+  if (lowerBoundInt && upperBoundInt) {
+    return success(size >= lowerBoundInt.getInt() &&
+                   size <= upperBoundInt.getInt());
+  }
+  if (lowerBoundInt) {
+    return success(size >= lowerBoundInt.getInt());
+  }
+  if (upperBoundInt) {
+    return success(size <= upperBoundInt.getInt());
+  }
+  return failure();
+}
```

Comment:
The existing code for dim bounds and divisibility uses the value bounds interface to support dynamic shapes too: https://github.com/iree-org/iree/blob/ce1445b05f3578b2629c08a9c31e04224dfc09ac/compiler/src/iree/compiler/Preprocessing/TransformExtensions/PreprocessingExtensions.cpp#L280

ValueBoundsInterface isn't great, we'd rather use integer range divisibility, but PDL patterns can't load an analysis so value bounds is the best alternative. While porting we need to keep support for the dynamic behavior since that's how specialization works.